sdks/java/harness/src/main/java/org/apache/beam/fn/harness/SplittableSplitAndSizeRestrictionsDoFnRunner.java

```diff
@@ -0,0 +1,490 @@
+      FnApiStateAccessor<Object> stateAccessor)
+      throws IOException {
+    this.pipelineOptions = pipelineOptions;
+
+    ParDoPayload parDoPayload = ParDoPayload.parseFrom(pTransform.getSpec().getPayload());
+
+    // DoFn and metadata
+    DoFn<InputT, OutputT> doFn = (DoFn<InputT, OutputT>) ParDoTranslation.getDoFn(parDoPayload);
+    DoFnSignature doFnSignature = DoFnSignatures.signatureForDoFn(doFn);
+    this.doFnInvoker = DoFnInvokers.tryInvokeSetupFor(doFn, pipelineOptions);
+    this.doFnSchemaInformation = ParDoTranslation.getSchemaInformation(parDoPayload);
+
+    // Main output
+    checkArgument(
+        pTransform.getOutputsMap().size() == 1,
+        "SplitAndSizeRestrictions expects exact one output, but got: ",
+        pTransform.getOutputsMap().size());
+    TupleTag<KV<KV<InputT, KV<RestrictionT, WatermarkEstimatorStateT>>, Double>> mainOutputTag =
+        new TupleTag<>(Iterables.getOnlyElement(pTransform.getOutputsMap().keySet()));
+    @SuppressWarnings("rawtypes") // cannot do this multi-level cast without rawtypes
+    FnDataReceiver<
+            WindowedValue<KV<KV<InputT, KV<RestrictionT, WatermarkEstimatorStateT>>, Double>>>
+        mainOutputConsumer =
+            (FnDataReceiver)
+                getPCollectionConsumer.apply(pTransform.getOutputsOrThrow(mainOutputTag.getId()));
+    this.mainOutputConsumer = mainOutputConsumer;
+
+    // Side inputs
+    this.sideInputMapping = ParDoTranslation.getSideInputMapping(parDoPayload);
+
+    // Register processing methods
+    this.observesWindow =
+        (doFnSignature.splitRestriction() != null
+                && doFnSignature.splitRestriction().observesWindow())
+            || (doFnSignature.newTracker() != null && doFnSignature.newTracker().observesWindow())
+            || (doFnSignature.getSize() != null && doFnSignature.getSize().observesWindow())
+            || !sideInputMapping.isEmpty();
+
+    if (observesWindow) {
+      this.mutableArgumentProvider = new SizedRestrictionWindowObservingArgumentProvider();
+    } else {
+      this.mutableArgumentProvider = new SizedRestrictionNonWindowObservingArgumentProvider();
+    }
+
+    this.stateAccessor = stateAccessor;
+  }
+
+  @Override
+  public @Nullable Object getCurrentKey() {
+    return null;
+  }
+
+  @Override
+  public BoundedWindow getCurrentWindow() {
+    return checkStateNotNull(
+        this.currentWindow,
+        "Attempt to access window outside windowed element processing context.");
+  }
+
+  private WindowedValue<InputT> getCurrentElement() {
+    return checkStateNotNull(
+        this.currentElement, "Attempt to access element outside element processing context.");
+  }
+
+  private Object getCurrentRestriction() {
+    return checkStateNotNull(
+        this.currentRestriction,
+        "Attempt to access restriction outside element processing context.");
+  }
+
+  private WatermarkEstimatorStateT getCurrentWatermarkEstimatorState() {
+    return checkStateNotNull(
```

Comment:
This check is failing. From what I can tell, all valid DoFns have a non-nullable state type. Pretty much all `Instant` and all derived from element watermarks, except one example that is meant to show off how flexible it is. So I don't know if it was already null or I introduced it. We didn't have any test failures prior to me adding the check, so it may have been a benign null passing through all the time. However, I expect I introduced it since this should always be a timestamp, basically.

Reply:
Confirmed that this is populated with a method that returns null if it is not implemented by the `DoFn`: https://github.com/apache/beam/blob/d5e3766091cd7da4c5157ea2c8242558243d162e/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/ByteBuddyDoFnInvokerFactory.java#L380

Reply:
It is an unsolvable typing problem, unless we redesign. I've narrowed the suppression to one method.